deepmd/pt/utils/auto_batch_size.py

```diff
@@ -78,26 +104,44 @@ def execute_with_batch_size(
             )
 
         index = 0
-        results = []
+        results = None
         while index < total_size:
             n_batch, result = self.execute(execute_with_batch_size, index, natoms)
-            if not isinstance(result, tuple):
-                result = (result,)
+            if not self.returned_dict:
+                result = (result,) if not isinstance(result, tuple) else result
             index += n_batch
-            if n_batch:
-                for rr in result:
-                    rr.reshape((n_batch, -1))
-                results.append(result)
-        r_list = []
-        for r in zip(*results):
+
+            def append_to_list(res_list, res):
+                if n_batch:
+                    res_list.append(res)
+                return res_list
+
+            if not self.returned_dict:
+                results = [] if results is None else results
+                results = append_to_list(results, result)
+            else:
+                results = (
+                    {kk: [] for kk in result.keys()} if results is None else results
+                )
+                results = {
+                    kk: append_to_list(results[kk], result[kk]) for kk in result.keys()
+                }
+
+        def concate_result(r):
             if isinstance(r[0], np.ndarray):
-                r_list.append(np.concatenate(r, axis=0))
+                ret = np.concatenate(r, axis=0)
             elif isinstance(r[0], torch.Tensor):
-                r_list.append(torch.cat(r, dim=0))
+                ret = torch.cat(r, dim=0)
             else:
                 raise RuntimeError(f"Unexpected result type {type(r[0])}")
-        r = tuple(r_list)
-        if len(r) == 1:
-            # avoid returning tuple if callable doesn't return tuple
-            r = r[0]
+            return ret
+
+        if not self.returned_dict:
```

Comment:
I think we don't need this argument? Just check the type of results

Reply:
good idea plz check [3c4ff2c](https://github.com/deepmodeling/deepmd-kit/pull/3626/commits/3c4ff2c1c34675c93b853902fd6b0d699034479a)